### ISSUE TITLE
SkillOpt: make human feedback target explicit and fix duplicate gate-retry behavior

### ISSUE THREAD

**Issue by jerryfane** [gitmoot/gitmoot#150]
## Summary

The current SkillOpt loop can mis-score optimizer candidates because imported human review feedback is not represented with a clear target. In the landing-page run, the feedback on `gitmoot-previews#21` was feedback about the previously reviewed baseline outputs, but the evaluator treated the current baseline as a fresh candidate that could be scored highly. That made the baseline look good even though the human feedback explicitly said the reviewed outputs were poor and should be refined.

We need to make the flow explicit:

- human review feedback targets the reviewed baseline outputs, not the new candidate;
- the optimizer candidate should be evaluated by whether it resolves the human feedback;
- gate retries should force a materially different patch when the previous retry produced a duplicate candidate;
- no-candidate diagnostics should explain this relationship clearly.

## Problem

In the latest run, Gitmoot correctly used the latest alpha versions and imported issue feedback, but the result was still confusing:

- baseline scored too high even though the imported human feedback said `quality: poor`, `continue_mode: refine`, `promote: no`;
- candidate lost narrowly against baseline;
- gate retry ran, but retry attempts produced the exact same candidate skill content;
- duplicate detection stopped the run with `repeated_duplicate_candidate`;
- the status did not clearly explain that the human feedback was about the baseline/reviewed outputs, not the candidate.

This creates the wrong mental model. The evaluator must not treat old human feedback as a direct judgment of the unseen candidate, but it also must not ignore that the baseline/reviewed outputs were explicitly rejected by the human.

## Desired Flow

```mermaid
flowchart TD
    A[Human reviews generated preview items] --> B[GitHub issue feedback YAML]
    B --> C[Gitmoot imports review feedback]
    C --> D[Create structured feedback packet]
    D --> E[feedback_target = baseline_review_outputs]
    E --> F[Optimizer updates skill from human feedback]
    F --> G[Target generates candidate outputs with updated skill]
    G --> H[Judge compares baseline issues vs candidate resolution]
    H --> I{Candidate materially resolves feedback?}
    I -- yes --> J[Candidate wins gate]
    J --> K[Run final checks/package candidate]
    I -- no --> L[Structured gate rejection]
    L --> M{Retryable and budget remains?}
    M -- yes --> N[Retry optimizer with anti-repeat context]
    N --> F
    M -- no --> O[Stop with clear no-candidate diagnostics]
```

### 1. Add Explicit Feedback Target Metadata

When feedback is imported from a review issue, store and propagate metadata such as:

```yaml
feedback_source: imported_human_review
feedback_target: baseline_review_outputs
review_issue: jerryfane/gitmoot-previews#21
review_run_id: landing-page-preview-trial-005-review-004
reviewed_skill_version: landing-page-builder@v12
```

The key point is that review feedback should be interpreted as:

> The human rejected or refined the reviewed baseline outputs. A new candidate must show that it resolves those issues.

It should not be interpreted as:

> The human already evaluated this unseen candidate.

### 2. Preserve Full Human Feedback Themes

The imported feedback currently appears too lossy. In the observed run, extracted themes were weak, mostly option labels such as `a`, `b`, `c`, `d`.

We should preserve structured human themes from the full text, for example:

```yaml
human_feedback_themes:
  - premium MoonAI-like branding direction
  - dark high-contrast visual identity
  - stronger product-relevant graphics/images
  - less generic SaaS layout
  - better proof sections and trust logos
  - stronger motion and scroll interactions
  - improved mobile responsiveness
  - clearer CTA and footer quality
```

These themes should be injected into optimizer and evaluator prompts as first-class data, not collapsed into ranking labels.

### 3. Use Comparative Resolution Scoring

The judge should compare candidate output against the baseline feedback packet, not score both as independent fresh outputs.

Recommended structured judge output:

```yaml
baseline_known_issues:
  - weak memorability
  - generic hero visual
  - insufficient product imagery
  - weak motion
  - mobile layout issues
candidate_resolution:
  branding: improved | unchanged | regressed
  imagery: improved | unchanged | regressed
  motion: improved | unchanged | regressed
  mobile: improved | unchanged | regressed
  proof_and_trust: improved | unchanged | regressed
selection_decision: candidate_wins | baseline_wins | reject_candidate
score_delta_reason: "Candidate changed spacing and CTA guidance but did not materially improve imagery or brand distinctiveness."
optimizer_hint: "Focus the next patch on product-relevant visuals, brand system, proof sections, motion, and mobile layout standards. Do not repeat the same CTA/footer-only patch."
```

This keeps the evaluator grounded in the actual optimization objective: resolving human feedback.

### 4. Gate Rule Should Prefer Material Feedback Resolution

The gate should not let generic completeness dominate when the human review is explicitly asking for refinement.

Generic completeness can still matter:

- valid artifact produced;
- page builds;
- hero/CTA/footer exists;
- mobile does not catastrophically break.

But candidate selection should prioritize whether the candidate materially addresses the imported feedback themes.

Example gate logic:

```text
candidate wins only if:
  hard requirements pass
  and candidate_resolution_score > baseline_resolution_score
  and at least one high-priority feedback theme materially improves
```

### 5. Fix Duplicate Gate Retry Behavior

Gate retry currently runs but can produce the exact same candidate file. When this happens, Gitmoot should not keep silently repeating the same patch.

Expected behavior:

```mermaid
flowchart TD
    A[Gate rejects candidate] --> B[Retry optimizer]
    B --> C{Candidate hash equals previous?}
    C -- no --> D[Evaluate new candidate]
    C -- yes --> E[Create duplicate retry packet]
    E --> F[Force anti-duplicate retry once]
    F --> G{Still duplicate?}
    G -- no --> D
    G -- yes --> H[Stop with repeated duplicate candidate diagnostics]
```

Duplicate retry packet should include:

```yaml
duplicate_candidate: true
previous_candidate_hash: ea03c9fa...
repeated_patch_summary:
  - strengthened human feedback instruction
  - added final CTA/footer guidance
  - added branding/mobile checks
anti_repeat_instruction: >
  The previous retry produced identical skill content. Do not repeat the same patch direction.
  Make a materially different update focused on unresolved human feedback themes: MoonAI-like premium branding,
  product-relevant imagery, motion, proof/trust sections, mobile layout quality, and memorability.
```

If the retry still duplicates, stop clearly.

### 6. Improve No-Candidate Diagnostics

No-candidate status should explain:

- which feedback issue was used;
- what the feedback targeted;
- whether baseline and candidate were scored as baseline-review resolution or fresh outputs;
- whether retries ran;
- whether retry candidates were duplicates;
- why promotion did not happen;
- what the next user action should be.

Example:

```yaml
no_candidate_reason: gate_rejected_best_origin_initial_skill
feedback_target: baseline_review_outputs
review_issue: jerryfane/gitmoot-previews#21
baseline_gate_score: 0.955
candidate_gate_score: 0.95
selection_result: candidate_lost
retry_attempts: 2/3
duplicate_retry_detected: true
stop_reason: repeated_duplicate_candidate
why_stopped: >
  The optimizer retried after gate rejection, but retry attempts produced identical candidate skill content.
next_action: >
  Rerun after improving optimizer retry context, or provide more specific feedback focused on the unresolved dimensions.
```

## Why This Design

This matches established evaluation and retry patterns:

- Evals should be structured tests with interpretable outputs, not only numeric scores.
- Graders should provide reasoning/evidence that can be reused for improvement.
- Prompt/system optimization should iterate from eval and human preference signals.
- Experiment analysis should compare candidate behavior against a baseline with score deltas and evaluator details.
- Retry systems should use bounded retries and classify retryable vs non-retryable failures.

## Acceptance Criteria

- Imported review feedback records `feedback_target: baseline_review_outputs`.
- Optimizer prompts receive full human feedback themes and review context.
- Evaluator prompts distinguish old reviewed outputs from new candidate outputs.
- Candidate selection is based on feedback resolution, not generic completeness alone.
- Gate rejection retry includes evaluator reason, score deltas, previous patch summary, and anti-repeat instruction.
- Duplicate candidate retry is detected by content hash.
- One anti-duplicate retry is attempted before stopping.
- No-candidate status clearly explains feedback target, score relation, retry attempts, duplicate detection, and next action.
- Regression tests cover:
  - poor/refine/promote=no feedback does not automatically veto unseen candidate;
  - poor/refine/promote=no feedback does mark baseline reviewed outputs as needing improvement;
  - candidate wins when it materially resolves imported feedback;
  - duplicate retry candidates trigger anti-repeat retry context;
  - repeated duplicate retry stops with clear diagnostics.

**Comment by jerryfane** [comment]
Implemented and merged across Gitmoot + gitmoot-skillopt.

PRs:
- jerryfane/gitmoot#151 - add feedback target metadata to exported training packages
- jerryfane/gitmoot-skillopt#52 - preserve package feedback context and full human feedback themes through the dataloader/optimizer path
- jerryfane/gitmoot-skillopt#53 - make evaluator scoring comparative against baseline review outputs instead of treating old human feedback as candidate feedback
- jerryfane/gitmoot-skillopt#54 - allow bounded gate retry for hard=0 human-feedback alignment cases and force stronger duplicate retry context
- jerryfane/gitmoot#152 - surface feedback target, review issue/run, score basis, score components, retry budget, and context-only diagnostics in CLI status

Verification:
- gitmoot-skillopt targeted pytest suites passed for dataloader, evaluator, and gate retry behavior.
- codex exec review --uncommitted passed cleanly for each PR after fixes.
- gitmoot Go tests could not run in this environment because go.mod requires Go 1.26 and the toolchain download is unavailable here; formatting and diff checks passed, and review passed cleanly.

**Comment by jerryfane** [comment]
Closed after merging PRs #151, #152, and gitmoot-skillopt PRs #52, #53, #54.
